caffe2/core/dispatch/DispatchKey.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#include "caffe2/core/dispatch/DeviceId.h"
+#include "caffe2/core/dispatch/LayoutId.h"
+#include "caffe2/core/typeid.h"
+
+#include <vector>
+#include <functional>
+#include "caffe2/utils/Array.h"
+
+namespace c10 {
+
+namespace details {
+struct TensorParameterDispatchKey final {
+  DeviceTypeId deviceTypeId;
+  LayoutId layoutId;
+  // TODO Move this CaffeTypeId to c10 namespace
+  caffe2::CaffeTypeId dataType;
+};
+inline constexpr bool operator==(const TensorParameterDispatchKey& lhs, const TensorParameterDispatchKey& rhs) {
+  return lhs.deviceTypeId == rhs.deviceTypeId && lhs.layoutId == rhs.layoutId && lhs.dataType == rhs.dataType;
+}
+}  // namespace details
+}  // namespace c10
+
+namespace std {
+  template<>
+  struct hash<c10::details::TensorParameterDispatchKey> {
+    // TODO constexpr hashing
+    size_t operator()(const c10::details::TensorParameterDispatchKey& obj) const {
+      return std::hash<c10::DeviceTypeId>()(obj.deviceTypeId) ^ std::hash<c10::LayoutId>()(obj.layoutId) ^ std::hash<caffe2::CaffeTypeId>()(obj.dataType);
+    }
+  };
+}  // namespace std
+
+namespace c10 {
+/**
+ * The dispatch key encodes the runtime type identity of a function call arguments,
+ * specifying what aspects of this identity can be dynamically dispatched on.
+ *
+ * Intuitively, given a function signature like f(Tensor, int), a valid dispatch
+ * key for the arguments might be [CPUFloatTensor] (notice that 'f' is NOT included
+ * in the dispatch key, and the runtime type of 'int' is NOT considered for dispatch
+ * (since it is trivial).
+ *
+ * Dispatch keys permit equality tests and are hashable.
+ *
+ * @tparam num_dispatch_args The number of dispatchable arguments
```

Comment:
IIRC, @zdevito proposed today that we not turn on multiple dispatch unconditionally for all operators; instead, we only apply it for the dozen or so ops that actually need multiple dispatch. It would be nice for this axis of the design space to be pointed out here, because this class, and `getTensorTypeIds_`, strongly imply that this design is NOT what is being planned for.

Reply:
This is a multi-dispatch-by-default design. We can rehash this discussion if we want to, but I don't think we converged on saying we do single-dispatch-by-default.